spec/lib/saml_idp/request_spec.rb

```diff
@@ -117,5 +121,67 @@ module SamlIdp
         expect(subject.response_url).to eq(subject.logout_url)
       end
     end
+
+    describe '#requested_aal_authn_context' do
+      subject { described_class.new raw_authn_request }
+
+      context "no aal context requested" do
+        let(:authn_context_classref) { '' }
+
+        it "should return nil" do
+          expect(subject.requested_aal_authn_context).to be_nil
+        end
+      end
+
+      context "only context requested is aal" do
+        let(:authn_context_classref) { build_authn_context_classref(aal) }
+
+        it "should return the aal uri" do
+          expect(subject.requested_aal_authn_context).to eq(aal)
+        end
+      end
+
+      context "multiple contexts requested including aal" do
+        let(:authn_context_classref) { build_authn_context_classref([ial, aal]) }
+
+        it "should return the aal uri" do
+          expect(subject.requested_aal_authn_context).to eq(aal)
+        end
+      end
+    end
+
+    describe '#requested_ial_authn_context' do
+      subject { described_class.new raw_authn_request }
+
+      context "no ial context requested" do
+        let(:authn_context_classref) { '' }
+
+        it "should return nil" do
+          expect(subject.requested_ial_authn_context).to be_nil
+        end
+      end
+
+      context "only context requested is ial" do
+        let(:authn_context_classref) { build_authn_context_classref(ial) }
+
+        it "should return the ial uri" do
+          expect(subject.requested_ial_authn_context).to eq(ial)
+        end
+      end
+
+      context "multiple contexts requested including ial" do
+        let(:authn_context_classref) { build_authn_context_classref([aal, ial]) }
+
+        it "should return the ial uri" do
+          expect(subject.requested_ial_authn_context).to eq(ial)
+        end
+      end
+    end
+
+    def build_authn_context_classref(contexts)
+      [contexts].flatten.map do |c|
```

Comment:
nit, not a huge deal: I would do `Array(contexts).map` because it handles this nicely and you don't need to flatten

```ruby
Array([1])
=> [1]
Array(1)
=> [1]
```